File: C#/300. Longest Increasing Subsequence.cs
/*
 * Link: https://leetcode.com/problems/longest-increasing-subsequence/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    public int LengthOfLIS(int[] nums) {
        var tails = new int[nums.Length];
        var size = 0;
        
        foreach(var n in nums) {
            var index = BinarySearch(tails, 0, size - 1, n);
            
            tails[index] = n;
            if (index == size) size++;
        }
        
        return size;
    }
    
    public int BinarySearch(int[] tails, int left, int right, int target) {
        if (left > right) return 0;
        
        while(left < right) {
            var mid = (left + right) / 2;
            
            if (tails[mid] < target) left = mid + 1;
            else right = mid;
        }
        
        if (tails[left] < target) return left + 1;
        return left;
    }
}

File: Python/300. Longest Increasing Subsequence.py
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        def binarySearch(arr, n):
            if len(arr) == 0: return 0
            
            left, right = 0, len(arr) - 1
            while(left < right):
                mid = (left + right) / 2
                if arr[mid] < n: 
                    left = mid + 1
                else: 
                    right = mid
            if arr[left] < n: 
                return left + 1
            return left
        res, ends = 0, []
        for n in nums:
            index = binarySearch(ends, n)
            res = max(res, index + 1)
            if index >= len(ends): 
                ends.append(n)
            else:
                ends[index] = n
        return res
            
            
        